components/region_cache_memory_engine/src/engine.rs

```diff
@@ -97,95 +133,98 @@ impl SnapshotList {
-    can_read: bool,
-    // Request with read_ts below it is not eligible for granting snapshot.
-    // Note: different region can have different safe_ts.
-    safe_ts: u64,
-}
-
-impl RegionMemoryMeta {
-    pub fn set_can_read(&mut self, can_read: bool) {
-        self.can_read = can_read;
+    pub(crate) fn is_empty(&self) -> bool {
+        self.0.is_empty()
     }
 
-    pub fn set_safe_ts(&mut self, safe_ts: u64) {
-        self.safe_ts = safe_ts;
+    pub(crate) fn len(&self) -> usize {
+        self.0.keys().len()
     }
 }
 
-#[derive(Default)]
-pub struct RegionCacheMemoryEngineCore {
-    engine: HashMap<u64, RegionMemoryEngine>,
-    region_metas: HashMap<u64, RegionMemoryMeta>,
+pub struct RangeCacheMemoryEngineCore {
+    engine: SkiplistEngine,
+    range_manager: RangeManager,
 }
 
-impl RegionCacheMemoryEngineCore {
-    pub fn mut_region_meta(&mut self, region_id: u64) -> Option<&mut RegionMemoryMeta> {
-        self.region_metas.get_mut(&region_id)
+impl RangeCacheMemoryEngineCore {
+    pub fn new(limiter: Arc<GlobalMemoryLimiter>) -> RangeCacheMemoryEngineCore {
+        RangeCacheMemoryEngineCore {
+            engine: SkiplistEngine::new(limiter),
+            range_manager: RangeManager::default(),
+        }
+    }
+
+    pub fn range_manager(&self) -> &RangeManager {
+        &self.range_manager
+    }
+
+    pub fn mut_range_manager(&mut self) -> &mut RangeManager {
+        &mut self.range_manager
     }
 }
 
-/// The RegionCacheMemoryEngine serves as a region cache, storing hot regions in
+/// The RangeCaheMemoryEngine serves as a range cache, storing hot ranges in
 /// the leaders' store. Incoming writes that are written to disk engine (now,
-/// RocksDB) are also written to the RegionCacheMemoryEngine, leading to a
-/// mirrored data set in the cached regions with the disk engine.
+/// RocksDB) are also written to the RangeCaheMemoryEngine, leading to a
+/// mirrored data set in the cached ranges with the disk engine.
 ///
-/// A load/evict unit manages the memory, deciding which regions should be
-/// evicted when the memory used by the RegionCacheMemoryEngine reaches a
-/// certain limit, and determining which regions should be loaded when there is
+/// A load/evict unit manages the memory, deciding which ranges should be
+/// evicted when the memory used by the RangeCaheMemoryEngine reaches a
+/// certain limit, and determining which ranges should be loaded when there is
 /// spare memory capacity.
 ///
-/// The safe point lifetime differs between RegionCacheMemoryEngine and the disk
-/// engine, often being much shorter in RegionCacheMemoryEngine. This means that
-/// RegionCacheMemoryEngine may filter out some keys that still exist in the
+/// The safe point lifetime differs between RangeCaheMemoryEngine and the disk
+/// engine, often being much shorter in RangeCaheMemoryEngine. This means that
+/// RangeCaheMemoryEngine may filter out some keys that still exist in the
```

Comment:
```suggestion
/// RangeCacheMemoryEngine may filter out some keys that still exist in the
```

Reply:
More instances of that below. Looks like a find/replace had a typo. Will let you fix the rest.